### PR TITLE
proposal: Add scopes to gitserver clients

### PR DESCRIPTION
**Pull request by eseliger**
This PR proposes a new pattern for instantiating gitserver clients. When we instantiate a new gitserver client, we should pass in a scope, a description of the environment it's used in. When a client is passed down to an environment, we can augment the client with an additional scope.

What is this for?

Looking at Grafana charts for dotcom, we see that we make about 2000 requests per second to gitserver. We know what endpoints we're hitting, and what _container_ is making the request. 
In Sourcegraph, containers are not a great boundary for services though. Some components stretch across multiple containers, and one container runs many different components, for example our worker container. While there are probably at least 50 routines owned by various different teams in that container, our current metrics only tell us that worker is making a large amount of requests. But we don't know who to talk to about it, because, again, worker is basically every team.

With scopes, we get more fine-grained insights and can group the metric by (container, op, scope), to get full insight into what _component_ (not _container_) is talking to gitserver.

Screenshot 2023-10-03 at 22 39 01@2x

Happy to revert to something simpler or more/less fine-grained later, but for now this should give us some more indication of what endpoints we need to prioritize the most.

## Test plan

Tested locally with `sg start monitoring`.

### CODE REVIEW

File: internal/gitserver/client.go [hunk 1]
```diff
@@ -112,6 +114,7 @@ func NewTestClient(t testing.TB) TestClient {
 		logger:      logger,
 		httpClient:  http.DefaultClient,
 		HTTPLimiter: limiter.New(500),
+		scope:       "gitserver.test",
```
Comment (pjlast) [inline internal/gitserver/client.go:117]: Since we have access to `t.Name()` here, might as well use it?

**Review by pjlast — COMMENTED** [review]
Seems good!

File: internal/gitserver/client.go (hunk 1 above)
Comment (eseliger) [reply internal/gitserver/client.go:117]: yep can do that!

**Review by pjlast — APPROVED** [review]

File: internal/gitserver/client.go [hunk 2]
```diff
@@ -88,10 +88,12 @@ type ClientSource interface {
 }
 
 // NewClient returns a new gitserver.Client.
-func NewClient() Client {
+// See Client.Scoped() for info on scoped clients.
+func NewClient(scope string) Client {
 	logger := sglog.Scoped("GitserverClient")
```
Comment (bobheadxi) [inline internal/gitserver/client.go:93]: @eseliger should we just add this scope to the logger too, to annotate error messages?

File: internal/gitserver/client.go [hunk 3]
```diff
@@ -241,6 +247,25 @@ type clientImplementor struct {
 	subRepoPermsChecker authz.SubRepoPermissionChecker
 }
 
+func (c *clientImplementor) Scoped(scope string) Client {
+	return &clientImplementor{
+		logger:       c.logger,
```
Comment (bobheadxi) [inline internal/gitserver/client.go:252]: similarly here, it might be helpful to propagate the scope to the logger

**Pull request merged**